### PR TITLE
Update quantiles document in the index the document belongs to

### PR DESCRIPTION
**Pull request by przemekwitek**
Currently, the quantiles document is always indexed into the index pointed by an alias ".ml-state-write".
We plan this alias->index mapping to change over time (as we introduce rollover).
Therefore, we need to make sure we don't end up with two copies of the quantiles document.
This PR achieves that by:
  1. searching for the doc
  2. either indexing it into the current write index or updating it in its current index

Relates #29938

### CODE REVIEW

**Comment by przemekwitek** [comment]
Some tests are failing. Let me debug that before reviewing.

**Comment by benwtrent** [comment]
Don't we write state documents much more often than we read them? 

I think we should handle multiple concrete indices on read instead of attempting to do so on write. 

What are the reasons for handling this type of situation on write instead of on read?

**Comment by droberts195** [comment]
> What are the reasons for handling this type of situation on write instead of on read?

In other indices where the same document ID could exist in multiple indices there is some mapped field that indicates which one is most recent.  So we can search for the document across the whole pattern, sort by the field that indicates the most recent and use size=1 to just get that.

With the `.ml-state` index there are no mappings, so we cannot easily sort as part of the search.

This is why I think the `.ml-state` index is different to the other indices where we deduplicate on read.

The other thing is that we don't write state documents _that_ often.  Nowhere near as frequently as results for example.

**Comment by benwtrent** [comment]
@droberts195 is the concern that choosing the state documents from the latest index won't always indicate the latest state document?

**Comment by droberts195** [comment]
There would have to be an ordering of the index names.  We know the most recent because that's what the write alias points to, but I don't think we easily know the ordering of the others unless we add some functionality to remember.  Some could get reindexed on major upgrades due to Lucene format changes.  I don't think doing this will reduce the amount of code.

**Comment by przemekwitek** [comment]
> Some tests are failing. Let me debug that before reviewing.

FYI: This is now fixed.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/persistence/JobResultsPersister.java [hunk 1]
```diff
@@ -255,17 +259,50 @@ public void persistCategoryDefinition(CategoryDefinition category, Supplier<Bool
...
      */
     public void persistQuantiles(Quantiles quantiles, Supplier<Boolean> shouldRetry) {
-        Persistable persistable = new Persistable(quantiles.getJobId(), quantiles, Quantiles.documentId(quantiles.getJobId()));
-        persistable.persist(AnomalyDetectorsIndex.jobStateIndexWriteAlias(), shouldRetry);
+        String quantilesDocId = Quantiles.documentId(quantiles.getJobId());
+
+        SearchRequest searchRequest =
+            new SearchRequest(AnomalyDetectorsIndex.jobStateIndexPattern())
+                .source(new SearchSourceBuilder().size(1).query(new IdsQueryBuilder().addIds(quantilesDocId)));
+        SearchResponse searchResponse = client.search(searchRequest).actionGet();
```
Comment (benwtrent) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/persistence/JobResultsPersister.java]: I am not 100% sure about this. 

This means that if we fail to search for the ID, the job will be flagged as failed immediately and stop running. 

I wonder if we should do retries on the synchronous search as we do for the persistence?

**Review by benwtrent — COMMENTED** [review]
Could you add a test where the max failure retries is 0 to verify that it is executed at least once?

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/utils/persistence/ResultsPersisterService.java [hunk 2]
```diff
@@ -85,29 +94,77 @@ public BulkResponse bulkIndexWithRetry(BulkRequest bulkRequest,
...
+        void nextIteration(String actionName, String failureMessage) {
+            currentAttempt++;
+
+            // If the outside conditions have changed and retries are no longer needed, do not retry.
             if (shouldRetry.get() == false) {
-                throw new ElasticsearchException("[{}] failed to index all results. {}", jobId, bulkResponse.buildFailureMessage());
+                String msg = new ParameterizedMessage(
+                    "[{}] should not retry {} after [{}] attempts. {}", jobId, actionName, currentAttempt, failureMessage)
+                    .getFormattedMessage();
+                LOGGER.info(() -> new ParameterizedMessage("[{}] {}", jobId, msg));
```
[the remark this answers is not in the document]
Comment (przemekwitek) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/utils/persistence/ResultsPersisterService.java]: Done.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/utils/persistence/ResultsPersisterService.java [hunk 3]
```diff
@@ -85,29 +94,77 @@ public BulkResponse bulkIndexWithRetry(BulkRequest bulkRequest,
...
+        }
+    }
+
+    public SearchResponse searchWithRetry(SearchRequest searchRequest,
+                                          String jobId,
+                                          Supplier<Boolean> shouldRetry,
+                                          Consumer<String> msgHandler) {
+        RetryContext retryContext = new RetryContext(jobId, shouldRetry, msgHandler);
+        while (true) {
+            SearchResponse searchResponse = client.search(searchRequest).actionGet();
```
[the remark this answers is not in the document]
Comment (przemekwitek) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/utils/persistence/ResultsPersisterService.java]: Done (catching `ElasticsearchException`).

**Review by przemekwitek — COMMENTED** [review]
> Could you add a test where the max failure retries is 0 to verify that it is executed at least once?

Done (`testSearchWithRetries_FailureAfterTooManyRetries_0`).

**Review by benwtrent — APPROVED** [review]

**Pull request merged**